### PR TITLE
Opening from script feature

### PR DESCRIPTION
**Pull request by Ivanluv**
This currently implements the opening of the split dialogue from script ,though not fully functional

To review this you need to :
1. Download this  file [RScript .zip](https://github.com/africanmathsinitiative/R-Instat/files/6106590/RScript.zip)
2. Extract the RScript .dll file and save  it to an appropriate folder on your system
3. Add the RScript .dll to your R-Instat project by 
  - opening the  `Solution Explorer`
 -  right click on `References` File
 -  select` Add Reference ...`
 -  select `Browse`
 -  navigate to the folder where you saved the .dll and select it

You can also view the source code for this RScript.dll file here https://github.com/lloyddewit/RScript

Among the things not functioning as required are:

- Incorrect updating of controls that use `AddAdditionalCodeParameterPair()`

### CODE REVIEW

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@Ivanluv This is just a partial review. I did not look at `clsRLink.vb` and `dlgSplitText.vb` yet.

File: instat/dlgSplitText.vb [hunk 1]
```diff
@@ -17,19 +17,19 @@
 
 Imports instat.Translations
 Public Class dlgSplitText
-    Public bFirstLoad As Boolean = True
+    Private bFirstLoad As Boolean = True
     Private bReset As Boolean = True
     Private clsTextComponentsFixed, clsTextComponentsMaximum As New RFunction
     Private clsBinaryColumns As New RFunction
-
+    Private lstRCodeStructure As List(Of RCodeStructure)
```
Comment (lloyddewit) [inline instat/dlgSplitText.vb:24]: We could make this a public property.
The `Set` could also set `bReset = True`.

File: instat/dlgSplitText.vb [hunk 2]
```diff
@@ -17,19 +17,19 @@
...
     Private Sub dlgSplitText_Load(sender As Object, e As EventArgs) Handles MyBase.Load
         autoTranslate(Me)
         If bFirstLoad Then
             InitialiseDialog()
             bFirstLoad = False
         End If
         If bReset Then
-            SetDefaults()
+            SetDefaults(lstRCodeStructure)
```
Comment (lloyddewit) [inline instat/dlgSplitText.vb:32]: This parameter is not needed

File: instat/dlgSplitText.vb [hunk 3]
```diff
@@ -87,36 +87,104 @@ Public Class dlgSplitText
         ucrSaveColumn.SetAssignToBooleans(bTempAssignToIsPrefix:=True)
     End Sub
 
-    Private Sub SetDefaults()
+    Private Sub SetDefaults(Optional lstOfRCodeStructure As List(Of RCodeStructure) = Nothing)
```
Comment (lloyddewit) [inline instat/dlgSplitText.vb:90]: Parameter is not needed

File: instat/dlgSplitText.vb [hunk 4]
```diff
@@ -87,36 +87,104 @@ Public Class dlgSplitText
...
+                                bMatchTwoRfunction = True
+                            End If
+                        End If
+                    End If
+                Next
+            Else
+                MsgBox("Developer error:The Script must be an RFunction")
+            End If
+            'This returns the list of Rcode from the script to nothing to prevent begguing when reseting
+            lstRCodeStructure = Nothing
```
Comment (lloyddewit) [inline instat/dlgSplitText.vb:162]: Move to outside of `if` statement?
Comment also needs fixing

File: instat/dlgSplitText.vb [hunk 5]
```diff
@@ -87,36 +87,104 @@ Public Class dlgSplitText
...
+        Else
+            lstRCodeStructure = Nothing
```
Comment (lloyddewit) [inline instat/dlgSplitText.vb:164]: Move to outside of `if` statement?

File: instat/dlgSplitText.vb [hunk 6]
```diff
@@ -87,36 +87,104 @@ Public Class dlgSplitText
...
+                                AddClsTextComponentsFixedDefaultParameters()
+
+                                ucrBase.clsRsyntax.SetBaseRFunction(clsBinaryColumns)
+                                bMatchTwoRfunction = True
+                            End If
+                        End If
+                    End If
+                Next
+            Else
+                MsgBox("Developer error:The Script must be an RFunction")
```
Comment (lloyddewit) [inline instat/dlgSplitText.vb:159]: If this line is reached then, after the message box, the dialog will be in an incomplete state. Better to restructure the `If` conditions above?

File: instat/dlgSplitText.vb [hunk 7]
```diff
@@ -87,36 +87,104 @@ Public Class dlgSplitText
...
+        If IsNothing(lstRCodeStructure) Then
+            'Continue with the updating Rfunction of Default value as normal
+            AddClsTextComponentsFixedDefaultParameters()
+
+            AddClsTextComponentsMaximumDefaultParameters()
+
+            ucrBase.clsRsyntax.SetBaseRFunction(clsTextComponentsFixed)
+            'This dialogue only requires one line of code to load hence this next elseIf statement
+        ElseIf (lstRCodeStructure.Count = 1) Then
+            If Not IsNothing(TryCast(lstOfRCodeStructure(0), RFunction)) Then
```
Comment (lloyddewit) [inline instat/dlgSplitText.vb:117]: Move all the `lstRCodeStructure` validity checks to the start of this function. If any of the checks fail then display a message box, reset `lstRCodeStructure` to `Nothing` and then continue.

File: instat/dlgSplitText.vb [hunk 8]
```diff
@@ -87,36 +87,104 @@ Public Class dlgSplitText
...
+                Dim strRcommand = TryCast(lstOfRCodeStructure(0), RFunction).strRCommand
+
+                For Each Rfunction In {clsTextComponentsFixed, clsTextComponentsMaximum, clsBinaryColumns}
+                    If strRcommand = Rfunction.strRCommand Then
+                        If bMatchTwoRfunction Then
+                            MsgBox("Developer error:More Than one Rfunctions match the Rfunction from the script")
+                            Exit For
+                            'TODO continue the normal loading of the dialogue if there
+                            'are more than one function matching the  Script
+                        Else
```
Comment (lloyddewit) [inline instat/dlgSplitText.vb:130]: Move all the `lstRCodeStructure` validity checks to the start of this function.

File: instat/dlgSplitText.vb [hunk 9]
```diff
@@ -87,36 +87,104 @@ Public Class dlgSplitText
...
+                            If Rfunction Is clsTextComponentsFixed Then
```
Comment (lloyddewit) [inline instat/dlgSplitText.vb:131]: Replace this `If ElseIf` structure with `Select Case`

File: instat/dlgSplitText.vb [hunk 10]
```diff
@@ -87,36 +87,104 @@ Public Class dlgSplitText
...
+            ucrBase.clsRsyntax.SetBaseRFunction(clsTextComponentsFixed)
+            'This dialogue only requires one line of code to load hence this next elseIf statement
+        ElseIf (lstRCodeStructure.Count = 1) Then
+            If Not IsNothing(TryCast(lstOfRCodeStructure(0), RFunction)) Then
+                'bMatchTwoRfunction Determines if the ListofRcodeStructure from the 
+                'script matches more than one RFunction
+                Dim bMatchTwoRfunction As Boolean = False
+                Dim strRcommand = TryCast(lstOfRCodeStructure(0), RFunction).strRCommand
+
+                For Each Rfunction In {clsTextComponentsFixed, clsTextComponentsMaximum, clsBinaryColumns}
```
Comment (lloyddewit) [inline instat/dlgSplitText.vb:123]: Rename `Rfunction` e.g. to `clsFunction`?

**Review by lloyddewit — CHANGES_REQUESTED** [review]

File: instat/dlgSplitText.vb [hunk 11]
```diff
@@ -87,36 +87,104 @@ Public Class dlgSplitText
...
+                                AddClsTextComponentsMaximumDefaultParameters()
+
+                                ucrBase.clsRsyntax.SetBaseRFunction(clsBinaryColumns)
+                                bMatchTwoRfunction = True
+                            ElseIf Rfunction Is clsTextComponentsMaximum Then
+                                clsTextComponentsMaximum = lstOfRCodeStructure(0)
+
+                                AddClsTextComponentsFixedDefaultParameters()
+
+                                ucrBase.clsRsyntax.SetBaseRFunction(clsBinaryColumns)
```
Comment (lloyddewit) [inline instat/dlgSplitText.vb:152]: Bug here?
```suggestion
                                ucrBase.clsRsyntax.SetBaseRFunction(clsTextComponentsMaximum)
```

File: instat/instat.vbproj [hunk 12]
```diff
@@ -128,6 +128,10 @@
     <Reference Include="RDotNet, Version=1.8.2.0, Culture=neutral, processorArchitecture=MSIL">
       <HintPath>..\packages\R.NET.1.8.2\lib\netstandard2.0\RDotNet.dll</HintPath>
     </Reference>
+    <Reference Include="RScript, Version=1.0.0.0, Culture=neutral, processorArchitecture=MSIL">
+      <SpecificVersion>False</SpecificVersion>
+      <HintPath>..\..\..\..\Pictures\RScript.dll</HintPath>
```
[the remark this answers is not in the document]
Comment (Ivanluv) [reply instat/instat.vbproj]: I would prefer adding in in the  ..\Source\Repos\R-Instat\packages

File: instat/ucrInputComboBox.vb [hunk 13]
```diff
@@ -198,8 +200,9 @@ Public Class ucrInputComboBox
         End If
     End Sub
 
-    Public Sub SetItems(dctItemParameterValuePairs As Dictionary(Of String, String), Optional bClearExisting As Boolean = True, Optional bSetCondtions As Boolean = True)
+    Public Sub SetItems(dctItemParameterValuePairs As Dictionary(Of String, String), Optional bClearExisting As Boolean = True, Optional bSetCondtions As Boolean = True, Optional bNewHasRParamFunctionAsString As Boolean = False)
```
[the remark this answers is not in the document]
Comment (Ivanluv) [reply instat/ucrInputComboBox.vb:203]: I have fixed this in a new PR #6252

File: instat/ucrInputComboBox.vb [hunk 14]
```diff
@@ -338,4 +341,25 @@ Public Class ucrInputComboBox
...
+                If Not IsNothing(clsNewCodeStructure.GetParameter(lstAllRParameters(0).strArgumentName)) Then
+                    If clsNewCodeStructure.GetParameter(lstAllRParameters(0).strArgumentName).bIsFunction Then
+                        'check that script is not empty.Make sure you make script before changing the clsRCodeStructure
+                        clsNewCodeStructure.GetParameter(lstAllRParameters(0).strArgumentName).strArgumentValue = clsNewCodeStructure.GetParameter(lstAllRParameters(0).strArgumentName).clsArgumentCodeStructure.ToScript()
+                        clsNewCodeStructure.GetParameter(lstAllRParameters(0).strArgumentName).bIsFunction = False
+                        clsNewCodeStructure.GetParameter(lstAllRParameters(0).strArgumentName).bIsOperator = False
+                        clsNewCodeStructure.GetParameter(lstAllRParameters(0).strArgumentName).bIsString = True
+                        clsNewCodeStructure.GetParameter(lstAllRParameters(0).strArgumentName).clsArgumentCodeStructure = Nothing
+                    End If
+                End If
```
[the remark this answers is not in the document]
Comment (Ivanluv) [reply instat/ucrInputComboBox.vb]: this is now in PR #6252

File: instat/dlgSplitText.vb (hunk 1 above)
Comment (Ivanluv) [reply instat/dlgSplitText.vb:24]: Done

File: instat/dlgSplitText.vb [hunk 15]
```diff
@@ -163,4 +233,9 @@ Public Class dlgSplitText
     Private Sub Controls_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrInputPattern.ControlContentsChanged, ucrReceiverSplitTextColumn.ControlContentsChanged, ucrNudPieces.ControlContentsChanged, ucrSaveColumn.ControlContentsChanged, ucrPnlSplitText.ControlContentsChanged
         TestOKEnabled()
     End Sub
+
+    Public Sub OpenFromScript(lstNewRcodeStructure As List(Of RCodeStructure))
+        bReset = True
+        lstRCodeStructure = lstNewRcodeStructure
+    End Sub
```
[the remark this answers is not in the document]
Comment (Ivanluv) [reply instat/dlgSplitText.vb:240]: Done

File: instat/dlgSplitText.vb (hunk 3 above)
Comment (Ivanluv) [reply instat/dlgSplitText.vb:90]: Done

File: instat/dlgSplitText.vb (hunk 5 above)
Comment (Ivanluv) [reply instat/dlgSplitText.vb:164]: Done

File: instat/dlgSplitText.vb (hunk 9 above)
Comment (Ivanluv) [reply instat/dlgSplitText.vb:131]: Done

File: instat/dlgSplitText.vb (hunk 10 above)
Comment (Ivanluv) [reply instat/dlgSplitText.vb:123]: I have reimplemented that function

File: instat/dlgSplitText.vb (hunk 11 above)
Comment (Ivanluv) [reply instat/dlgSplitText.vb:152]: Thanks.I have changed this

File: instat/dlgSplitText.vb (hunk 6 above)
Comment (lloyddewit) [reply instat/dlgSplitText.vb:125]: Done, thanks

**Review by lloyddewit — CHANGES_REQUESTED** [review]

**Comment by lloyddewit** [comment]
@maxwellfundi This is a long-term, ongoing PR. @Ivanluv and I agreed that we will put this PR on hold during the sprint, and we will continue after 26 March 2021. I'll be happy to review this PR again after the sprint.

**Comment by maxwellfundi** [comment]
@lloyddewit  that sound okay, I will now change its milestone

**Pull request closed without merging**